presto-main/src/test/java/io/prestosql/type/TestArrayOperators.java

```diff
@@ -749,6 +750,23 @@ public void testSubscript()
         assertDecimalFunction("ARRAY [2.22222222222222222, 2.3][1]", decimal("2.22222222222222222"));
     }
 
+    @Test
+    public void testSubscriptReturnType()
+    {
+        // Test return type of array subscript by passing it to another operation
+        // One test for each specialization of the operator
+        assertSubscriptCast("true", "true"); // boolean
+        assertSubscriptCast("1234", "1234"); // long
+        assertSubscriptCast("1.23", "1.23"); // double
+        assertSubscriptCast("IPADDRESS '127.0.0.1'", "127.0.0.1"); // slice
+        assertSubscriptCast("TIMESTAMP '1970-01-01 00:00:00.000000001'", "1970-01-01 00:00:00.000000001"); // object (LongTimestamp)
```

Comment:
`select CAST((ARRAY [TIMESTAMP '1970-01-01 00:00:00.000000001'])[1] AS VARCHAR);` does not fail even without your fixes.

Reply:
It does in this test.

Reply:
Indeed :) WTF ...

Reply:
`assertFunction` executes the expression five different ways. At least one of them involves manually compiling the expression, which makes the bug occur.

It fails at `FunctionAssertions` line 502. But to get there, the query had to already succeed as-is on line 493.

Reply:
@losipiuk this query doesn't fail, as the subscript (and everything else) is taken care of by the expression interpreter, during optimization, doesn't reach compiled code at all

the TestArrayOperators exercises both optimized and non-optimized expressions.